src/crimson/os/seastore/onode_manager/staged-fltree/value.h

```diff
@@ -0,0 +1,277 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <ostream>
+
+#include "include/buffer.h"
+#include "crimson/common/type_helpers.h"
+
+#include "fwd.h"
+#include "node_extent_mutable.h"
+
+namespace crimson::os::seastore::onode {
+
+// value size up to 64 KiB
+using value_size_t = uint16_t;
+enum class value_magic_t : uint8_t {
+  TEST = 0x52,
+  ONODE,
+};
+inline std::ostream& operator<<(std::ostream& os, const value_magic_t& magic) {
+  switch (magic) {
+  case value_magic_t::TEST:
+    return os << "TEST";
+  case value_magic_t::ONODE:
+    return os << "ONODE";
+  default:
+    return os << "UNKNOWN(" << magic << ")";
+  }
+}
+
+/**
+ * value_config_t
+ *
+ * Parameters to create a value.
+ */
+struct value_config_t {
+  value_magic_t magic;
+  value_size_t payload_size;
+
+  value_size_t allocation_size() const;
+
+  void encode(ceph::bufferlist& encoded) const {
+    ceph::encode(magic, encoded);
+    ceph::encode(payload_size, encoded);
+  }
+
+  static value_config_t decode(ceph::bufferlist::const_iterator& delta) {
+    value_magic_t magic;
+    ceph::decode(magic, delta);
+    value_size_t payload_size;
+    ceph::decode(payload_size, delta);
+    return {magic, payload_size};
+  }
+};
+inline std::ostream& operator<<(std::ostream& os, const value_config_t& conf) {
+  return os << "ValueConf(" << conf.magic
+            << ", " << conf.payload_size << "B)";
+}
+
+/**
+ * value_header_t
+ *
+ * The header structure in value layout.
+ *
+ * Value layout:
+ *
+ * # <- alloc size -> #
+ * # header | payload #
+ */
+struct value_header_t {
+  value_magic_t magic;
+  value_size_t payload_size;
+
+  value_size_t allocation_size() const {
+    return payload_size + sizeof(value_header_t);
+  }
+
+  const char* get_payload() const {
+    return reinterpret_cast<const char*>(this) + sizeof(value_header_t);
+  }
+
+  NodeExtentMutable get_payload_mutable(NodeExtentMutable& node) const {
+    return node.get_mutable_absolute(get_payload(), payload_size);
+  }
+
+  char* get_payload() {
+    return reinterpret_cast<char*>(this) + sizeof(value_header_t);
+  }
+
+  void initiate(NodeExtentMutable& mut, const value_config_t& config) {
+    value_header_t header{config.magic, config.payload_size};
+    mut.copy_in_absolute(this, header);
+    mut.set_absolute(get_payload(), 0, config.payload_size);
```

Comment:
why do we need to initialize the payload beforehand?

Reply:
I think this is just simpler to return a new value with payload all initialized to 0.

Otherwise it will contain random garbage values when return back to user at the first time (at insertion), and user has to initialize the payload by themselves. Currently there is no user-defined initialization. The setter methods cannot be used for initialization because they are replayable.

Reply:
> and user has to initialize the payload by themselves

yes, i think they *have to* initialize the payload by themselves.

Reply:
Got your point. Working on it...